examples/gno.land/r/demo/pagination/pagination.gno

```diff
@@ -0,0 +1,137 @@
+// It keeps a slice of signed up addresses for efficient pagination
+package pagination
+
+import (
+	"std"
+	"strconv"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/mux"
+	"gno.land/p/demo/ufmt"
+)
+
+type Signup struct {
+	account string
+	height  int64
+}
+
+// signups - keep a slice of signed up addresses efficient pagination
+var signups []Signup
+
+// tracker - keep track of who signed up
+var tracker *avl.Tree
+
+const pageSize = 20
+
+func init() {
+	tracker = avl.NewTree()
+	SignUp()
+}
+
+func SignUp() string {
+	caller := std.GetOrigCaller().String()
+	height := std.GetHeight()
+
+	_, exists := tracker.Get(caller)
+	if exists {
+		panic(caller + " is already signed up!")
+	}
+
+	tracker.Set(caller, struct{}{})
+	signup := Signup{
+		caller,
+		height,
+	}
+
+	signups = append(signups, signup)
+	return ufmt.Sprintf("%s signed up at block #%d!", signup.account, signup.height)
+}
+
+func GetSignedUpInRange(page, pageSize int) []Signup {
+	if page < 1 {
+		panic("page number cannot be less than 1")
+	}
+
+	if pageSize < 1 || pageSize >= 50 {
+		panic("page size must be between 1 and 50")
+	}
+
+	// Pagination
+	// Calculate indexes
+	startIndex := (page - 1) * pageSize
+	endIndex := startIndex + pageSize
+
+	// If page does not contain any posts
+	if startIndex >= len(signups) {
+		return nil
+	}
+
+	// If page contains fewer posts than the page size
+	if endIndex > len(signups) {
+		endIndex = len(signups)
+	}
+
+	return signups[startIndex:endIndex]
+}
+
+func renderHelper(res *mux.ResponseWriter, req *mux.Request) {
+	totalSignups := len(signups)
+
+	res.Write("# Welcome to Gno.land!\n\n")
+	res.Write(ufmt.Sprintf("## Signed up users (%d):\n\n", totalSignups))
+
+	page, err := strconv.Atoi(req.GetVar("number"))
+	if err != nil {
+		page = 1 // render first page on bad input
+	}
+
+	if totalSignups == 0 {
+		res.Write("No users are signed up currently.")
+		return
+	}
+
+	fetchedSignups := GetSignedUpInRange(page, pageSize)
+	if totalSignups == 0 {
+		res.Write("No users on this page\n\n")
+	}
+
+	pageStartIndex := pageSize * (page - 1)
+	for i, signup := range fetchedSignups {
+		out := ufmt.Sprintf("#### User #%d - %s - signed up at Block #%d\n", pageStartIndex+i, signup.account, signup.height)
+		res.Write(out)
+	}
+
+	res.Write("---\n\n")
+
+	latestSignup := signups[totalSignups-1]
+	res.Write(ufmt.Sprintf("Latest signup: %s at Block #%d\n\n", latestSignup.account, latestSignup.height))
+
+	res.Write("---\n\n")
+
+	res.Write(ufmt.Sprintf("You're viewing page #%d\n\n", page))
+
+	var prevPage string
+	var nextPage string
+	if page < 2 {
+		prevPage = ""
+	} else {
+		prevPage = ufmt.Sprintf("[Previous page](/r/demo/pagination:page/%d)\n\n", page-1)
+	}
+
+	if len(fetchedSignups) >= pageSize {
+		nextPage = ufmt.Sprintf("[Next page](/r/demo/pagination:page/%d)\n\n", page+1)
+	} else {
+		nextPage = ""
+	}
+
+	res.Write(prevPage)
+	res.Write(nextPage)
+}
+
+func Render(path string) string {
+	router := mux.NewRouter()
```

Comment:
Does this router get created an configured for every call to `Render`? If so, it would avoid some unnecessary cycles if the router were a realm variable that is configured during realm initialization.

Reply:
Pretty sure it does. I'll try to figure out how many cycles it takes in both cases. Nice suggestion 👍

Reply:
From some light testing: 

```
func TestC(t *testing.T) {
	Render("")
}

// Mux in render
//--- PASS: TestC (0.00s)
//---       runtime: cycle=15.5k imports=6 allocs=444.7k(0.09%)

// Mux in Init
//--- PASS: TestC (0.00s)
//---       runtime: cycle=14.9k imports=6 allocs=433.8k(0.09%)
```

Nice! Thanks

Reply:
Added it [9e748e8](https://github.com/gnolang/gno/pull/1949/commits/9e748e83c8d00bb6a65c54adff8951f585707aef)